### PR TITLE
GRPO implementation

### PR DESCRIPTION
**Pull request by abaheti95**
Adds critic free model pathway to online RL code (for [GRPO](https://arxiv.org/abs/2402.03300)) and modifies some function signatures and code pathways.
- new `hf_critic_free_lm` model class similar to DPO HF model (i.e. doesn't have `config_overrides` but directly passes additional params via model_config)
- Includes `normalize_advantage` and `length_normalize_policy_loss` flags for critic-free model. Set these both to `false` for [Dr. GRPO objective](https://arxiv.org/abs/2503.20783). The `length_normalize_policy_loss` also modifies KL length normalization.
- `ppo_loss` --> `online_rl_loss`. Shared pathway for both ppo and grpo with `critic_free` boolean flag
- `composer_ppo_forward` --> `composer_online_rl_forward`. Shared pathway for both ppo and grpo with `critic_free` path
- Logs both `seq_kl` and `token_kl` of all variants in the loss logs

Working runs with new patch:
- GRPO `mcli logs test-critic-free-pr-grpo-kl-fix-CcXvGF -f`
- PPO `mcli logs test-critic-free-pr-ppo-kl-fix-sXRGBT -f`

NOTE: Need to modify the PPO model definition to nuke `config_overrides` in the future

Autoresumption also working with GRPO
image

### CODE REVIEW

File: compose_rl/ppo/callback.py [hunk 1]
```diff
@@ -269,9 +263,19 @@ def env_reward(
...
+
+            # Need to add in the padding for the value function
+            value_action_mask = torch.cat([
+                action_mask,
+                torch.zeros((batch_size, 1), device=cur_device),
+            ],
+                                          dim=-1)
+            device_train_microbatch_values *= value_action_mask
+            partial_env_output['values'
+                              ] = device_train_microbatch_values.detach()
```
Comment (bcui-db) [inline compose_rl/ppo/callback.py:278]:
```suggestion
                              ] = device_train_microbatch_values
```

Comment (abaheti95) [reply compose_rl/ppo/callback.py:278]: Cool. Removed all `detach()` from `partial_env_output` within `torch.no_grad()`

File: compose_rl/ppo/callback.py [hunk 2]
```diff
@@ -321,19 +325,44 @@ def __init__(
         # Value used in the generalized advantage estimate calculation.
         self.lambda_gae = var_config.get('lambda_gae', 1.0)
 
+        # Other algo specific hparams
```
[the remark this answers is not in the document]
Comment (abaheti95) [reply compose_rl/ppo/callback.py:327]: changed to `loss_type` as discussed offline

File: compose_rl/ppo/callback.py [hunk 3]
```diff
@@ -321,19 +324,46 @@ def __init__(
...
+        # Find if we are using a critic free model or not
+        if train_config['model']['name'] == 'hf_critic_free_lm':
+            self.loss_type = 'grpo'
```
Comment (bcui-db) [inline compose_rl/ppo/callback.py:330]: Can you make this a variable in `state.model` and we get it from there? So we can add other things like RLOO, REBEL, etc.

File: compose_rl/ppo/model.py [hunk 4]
```diff
@@ -226,3 +234,78 @@ def determine_early_stop(self):
...
+
+
+class ComposerHFCriticFreePolicyModel(ComposerHFCausalLM):
+    """HF class wrapper for Critic Free Policy model."""
+    default_train_metrics: tuple = ()
+    default_eval_metrics: tuple = ()
+
+    def __init__(
+        self,
+        normalize_advantage: bool = True,
```
Comment (bcui-db) [inline compose_rl/ppo/model.py:246]: Can you add a docstring here explaining the variables?

File: compose_rl/ppo/model.py [hunk 5]
```diff
@@ -226,3 +234,78 @@ def determine_early_stop(self):
...
+        self.compute_kl_loss = compute_kl_loss
+        self.target_kl = target_kl
+        self.kl_estimator = kl_estimator
+        self.kl_clip_range = kl_clip_range
+
+    def forward(self, batch: MutableMapping):
+        ret_val = composer_online_rl_forward(
+            batch,
+            self.model,
+            loss_type='grpo',
```
Comment (bcui-db) [inline compose_rl/ppo/model.py:272]: When you add the above loss_type, also modify it here

File: compose_rl/ppo/modeling_utils.py [hunk 6]
```diff
@@ -116,158 +117,270 @@ def composer_ppo_forward(
...
     kl_estimator: Optional[str] = 'k1',
     kl_clip_range: Optional[float] = 40.0,
 ) -> tuple[MutableMapping, torch.Tensor]:
-    """Compute the PPO loss.
+    """Compute the online RL loss.
 
     Args:
         outputs (MutableMapping): The outputs from the forward pass.
         batch (MutableMapping): The batch to compute the loss over.
+        loss_type (str): The loss type which decides whether to use critic-free or not. Defaults to "ppo".
```
Comment (bcui-db) [inline compose_rl/ppo/modeling_utils.py:154]:
```suggestion
        loss_type (str): The loss type which decides whether to use critic-free or not. Defaults: ``ppo``.
```

File: compose_rl/ppo/modeling_utils.py [hunk 7]
```diff
@@ -116,158 +117,270 @@ def composer_ppo_forward(
...
+        if adv_masked_var.dim() > 0:
+            adv_masked_var = adv_masked_var[0]
+
+        # Normalizing advantages over each minibatch
+        advantages = utils.masked_normalize(
+            batch['advantages'],
+            adv_masked_mean,
+            adv_masked_var,
+        )
+    elif loss_type == 'grpo':
```
Comment (bcui-db) [inline compose_rl/ppo/modeling_utils.py:222]: Can we do two things here: 

1. Initialize `values, v_preds, value_loss, value_clip_frac, returns_mean, returns_var, val_error` to `None` above the `ppo` conditional?
2. We should do the `loss_type` check just above and it be sth like `if loss_type` not in ["ppo", "grpo"] raise the ValueError.

File: compose_rl/ppo/modeling_utils.py [hunk 8]
```diff
@@ -116,158 +117,270 @@ def composer_ppo_forward(
...
+    else:
+        policy_loss = utils.masked_sum(policy_loss, batch['action_mask'])
+
+    # policy_kl_logging_dict = {
+    #     f'kl/policy_kl_{k}_estimate': v for k, v in policy_kl_dict.items()
+    # }
+    # online_ift_kl_logging_dict = {
+    #     f'kl/online_ift_kl_{k}_estimate': v
+    #     for k, v in online_ift_kl_dict.items()
+    # }
```
Comment (bcui-db) [inline compose_rl/ppo/modeling_utils.py:303]:
```suggestion
```

**Review by bcui-db — APPROVED** [review]
Minor comments, but overall LGTM.

**Comment by abaheti95** [comment]
Addressed all the comments. Also made minor changes to grpo advantages. Instead of forwarding the flat advantage tensor, it now keeps the same shape of advantages for both ppo and grpo.. In grpo, the flat advantage is copied across the action_mask instead. This reduces a few lines of code.
I re-ran the baseline PPO and GRPO runs again with this change, and both are working fine.

- `mcli logs test-critic-free-pr-grpo-kl-fix-DJ4yCU -f`
- `mcli logs test-critic-free-pr-ppo-kl-fix-5lFOaf -f`
